node/node.go

```diff
@@ -515,40 +491,76 @@ func (n *Node) Remotes() []api.Peer {
 	return remotes
 }
 
-func (n *Node) loadCertificates() error {
-	certDir := filepath.Join(n.config.StateDir, "certificates")
-	rootCA, err := ca.GetLocalRootCA(certDir)
-	if err != nil {
-		if err == ca.ErrNoLocalRootCA {
-			return nil
+func (n *Node) loadSecurityConfig(ctx context.Context) (*ca.SecurityConfig, error) {
+	paths := ca.NewConfigPaths(filepath.Join(n.config.StateDir, certDirectory))
+
+	// Check if we already have a CA certificate on disk.
+	rootCA, err := ca.GetLocalRootCA(paths.RootCA)
+	switch {
+	case err == nil:
+		log.G(ctx).Debug("loaded CA certificate")
+	case err == ca.ErrNoLocalRootCA && n.config.JoinAddr == "":
+		// if we're not joining a cluster, bootstrap a new one - and we have to set the unlock key
+		n.unlockKey = nil
+		if n.config.AutoLockManagers {
+			n.unlockKey = encryption.GenerateSecretKey()
+		}
+		rootCA, err = ca.CreateRootCA(ca.DefaultRootCN, paths.RootCA)
+		if err != nil {
+			return nil, err
+		}
+		log.G(ctx).Debug("generated CA key and certificate")
+	case err == ca.ErrNoLocalRootCA:
+		log.G(ctx).WithError(err).Debugf("failed to load local CA certificate")
+		rootCA, err = ca.DownloadRootCA(ctx, paths.RootCA, n.config.JoinToken, n.remotes)
+		if err != nil {
+			return nil, err
 		}
-		return err
 	}
-	configPaths := ca.NewConfigPaths(certDir)
-	clientTLSCreds, _, err := ca.LoadTLSCreds(rootCA, configPaths.Node)
-	if err != nil {
-		if os.IsNotExist(err) {
-			return nil
+
+	// Obtain new certs and setup TLS certificates renewal for this node:
+	// - We call LoadOrCreateSecurityConfig which blocks until a valid certificate has been issued
+	// - We retrieve the nodeID from LoadOrCreateSecurityConfig through the info channel. This allows
+	// us to display the ID before the certificate gets issued (for potential approval).
+	// - We wait for LoadOrCreateSecurityConfig to finish since we need a certificate to operate.
+	// - Given a valid certificate, spin a renewal go-routine that will ensure that certificates stay
+	// up to date.
+	issueResponseChan := make(chan api.IssueNodeCertificateResponse, 1)
+	go func() {
+		select {
+		case <-ctx.Done():
+		case resp := <-issueResponseChan:
+			log.G(log.WithModule(ctx, "tls")).WithFields(logrus.Fields{
+				"node.id": resp.NodeID,
+			}).Debugf("requesting certificate")
```

Comment:
Is this log message correct? It seems like this code would be reached after the new certificate has been issued.